### PR TITLE
FIX #239: Added support for animated dialogs

### LINKED ISSUE #239: `active` and `paused` props don't work
I have smoothly animating dialogs that fade in and fade out.

All the examples assume that the dialogs have no intro or outro animations.

I've forked the repository and converted the "demo-defaults" example into a test case example of what I need FocusTrap to be able to do:
https://github.com/Dan503/focus-trap-react/blob/master/demo/js/demo-defaults.js#L36

There are `active` and `paused` props according to the index.d.ts file but when I try to use them, I get this error:

```
Uncaught Error: Your focus-trap must have at least one container with at least one tabbable node in it at all times
```

Basically this is what I'm asking for (in pseudo code terms):

````
if ( `active` prop equals `true` ) {
    activate the focus-trap
}
else if ( `active` prop equals `false` ) {
    deactivate the focus-trap
}

if ( `paused` prop equals `true` ) {
    pause the focus-trap
}
else if ( `paused` prop equals `false` ) {
    resume the focus-trap
}
````

### PR DESCRIPTION
**Pull request by Dan503**
Fix for issue #239

Adds an animated example for testing.

Also added "View demo source" links to make it easier for people to learn how to use the component.

I attempted to create a cypress test for the new demo but I wasn't able to get Cypress working so I have no idea if the test is passing or not.

<details>
<summary>PR Checklist</summary>
<br/>

__Please leave this checklist in your PR.__

- Issue being fixed is referenced.
- Source changes maintain:
  - Stated browser compatibility.
  - Stated React compatibility.
- Unit test coverage added/updated.
- E2E test coverage added/updated.
- Prop-types added/updated.
- Typings added/updated.
- README updated (API changes, instructions, etc.).
- Changes to dependencies explained.
- Changeset added (run `yarn changeset` locally to add one, and follow the prompts).
  - EXCEPTION: A Changeset is not required if the change does not affect any of the source files that produce the package bundle. For example, tooling changes, test updates, or a new dev-only dependency to run tests more efficiently should not have a Changeset since it will not affect package consumers.

</details>

### CODE REVIEW

**Review by stefcameron — CHANGES_REQUESTED** [review]
@Dan503 Thanks for posting this PR! It's nice to have the "View source" links for all the samples. It's neat that you attempt to determine when the container is visible instead of asking for a simple time delay (but this does have negative consequences on performance).

Some thoughts:
- If `focus-trap` needs to solve this, we need a solution that doesn't rely on browser APIs that cause reflow.
- I think a solution to this problem is better directed to https://github.com/focus-trap/focus-trap than `focus-trap-react` because it's not React-specific problem.
- If all you need a delay before activating the trap, why is it not a viable option to use a local state variable on your "open dialog" button to simply delay calling `focusTrap.activate()` in your code by 0.6s in this case? The trap is not activated by default, so you have to already be actively activating it anyway. Just delay that by 0.6s...

File: src/focus-trap-react.js [hunk 1]
```diff
@@ -101,17 +114,25 @@ class FocusTrap extends React.Component {
...
-        }
+        const deactivate = () => {
+          // NOTE: we never let the trap return the focus since we do that ourselves
+          this.focusTrap.deactivate({ returnFocus: false });
+          if (this.returnFocusOnDeactivate) {
+            this.returnFocus();
+          }
+        };
+
+        deactivate();
```
[the remark this answers is not in the document]
Comment (Dan503) [reply src/focus-trap-react.js:125]: I originally also delayed the deactivation of the trap but then I realised that it wasn't necessary to delay it.

The `deactivate()` helps add a bit of clarity to the code in my opinion but doesn't change the functionality.

If you want this can be reverted.

**Comment by Dan503** [comment]
> If all you need a delay before activating the trap, why is it not a viable option to use a local state variable on your "open dialog" button to simply delay calling focusTrap.activate() in your code by 0.6s in this case? The trap is not activated by default, so you have to already be actively activating it anyway. Just delay that by 0.6s...

The animation looks best if the focus is applied as soon as the element is focusable. Otherwise you get this jarring experience where the animation ends and then you see the focus suddenly get applied. There is also probably an accessibility concern moving focus some time after the user activated a dialogue.

 That is why I checked for visibility with `getComputedStyle()`. I figured that since it was being used by the core focus-trap code it would be ok to use here.

Basically the problem that needs to be solved is that we don't want to activate the trap until the focus-trap becomes focusable. When there is an animation involved, there is a delay between the client code calling for activation and the focus-trap being focusable. The time this takes is variable and based on how fast the browser runs the animation. We don't know in js exactly when the element is focusable.

We could do a 5 millisecond delay perhaps with `setTimeout()` but that is just a guess and doesn't guarantee that the focus-trap is focusable. It would work for 90% though probably (that's a guess as well).

**Comment by Dan503** [comment]
> If all you need a delay before activating the trap, why is it not a viable option to use a local state variable on your "open dialog" button to simply delay calling focusTrap.activate() in your code by 0.6s in this case? The trap is not activated by default, so you have to already be actively activating it anyway. Just delay that by 0.6s...

The animation looks best if the focus is applied as soon as the element is focusable. Otherwise you get this jarring experience where the animation ends and then you see the focus suddenly get applied. There is also probably an accessibility concern moving focus some time after the user activated a dialogue.

 That is why I checked for visibility with `getComputedStyle()`. I figured that since it was being used by the core focus-trap code it would be ok to use here.

Basically the problem that needs to be solved is that we don't want to activate the trap until the focus-trap becomes focusable. When there is an animation involved, there is a delay between when the client code calls for activation and when the focus-trap is able to apply focus.

We could use a setTimeout of 5ms but that isn't a totally reliable thing. It's better if we activate the trap exactly when the trap is able to be activated.

**Comment by Dan503** [comment]
Sorry my browser bugged out when posting that last comment

**Comment by stefcameron** [comment]
> That is why I checked for visibility with `getComputedStyle()`. I figured that since it was being used by the core focus-trap code it would be ok to use here.

Makes sense. We do use it there. But we had to add this `displayCheck` feature so that others could _avoid_ using it due to performance issues, so I'd rather avoid adding another instance of it if at all possible.

> Basically the problem that needs to be solved is that we don't want to activate the trap until the focus-trap becomes focusable. When there is an animation involved, there is a delay between when the client code calls for activation and when the focus-trap is able to apply focus.
> 
> We could use a setTimeout of 5ms but that isn't a totally reliable thing. It's better if we activate the trap exactly when the trap is able to be activated.

How about a solution where the code executed to determine visibility is provided by the consumer? For example, `focus-trap` (not `focus-trap-react`) gets some new option where you give it a handler, and it will call this handler, which maybe returns true or false, to know when the trap should finally be activated. To use this from `focus-trap-react`, you provide that handler as part of the `focusTrapOptions` prop.

**Comment by Dan503** [comment]
> How about a solution where the code executed to determine visibility is provided by the consumer?

That sounds like a good solution. It also means we don't have to make assumptions about how the consumer is doing their animation then.

I'll try giving that a go.

The function can default to a function that just returns true immediately.

I'm a bit busy today so might not get around to doing this until next weekend.

**Comment by Dan503** [comment]
I've update this branch with a proof of concept for how it will roughly work when the code is moved across into the core focus-trap library.

I'll update the focus-trap library later.

**Comment by stefcameron** [comment]
> The function can default to a function that just returns true immediately.

By default, we'll want the current behavior to take place, so I would say if the option is not a function, the trap should activate immediately. If we default to a function that returns true and always wait for it before activating, we will introduce a next-frame activation for everyone, and that will break current behavior. 

> I'm a bit busy today so might not get around to doing this until next weekend.

No worries, no rush! Thanks for taking this on, I appreciate the help.
